### PR TITLE
[nfc][InstrProfTest]Factor out common code for value profile test

### PR DESCRIPTION
**Pull request by mingmingl-llvm**
Three existing test cases {get_icall_data_read_write, get_icall_data_read_write_with_weight, get_icall_data_read_write_big_endian} have common test data and testing logic. Extract common code into  `testICallDataReadWrite`.
- Add helper function `addValueProfData` and `testValueDataArray`. This two helper functions are used by `testICallDataReadWrite`, and possibly other test cases.

### CODE REVIEW

**Review by david-xl — COMMENTED** [review]
Are these changes inter-dependent or can be further split into smaller NFC patches?

**Comment by minglotus-6** [comment]
> Are these changes inter-dependent or can be further split into smaller NFC patches?

This could be further split. Keep the refactor of three related tests in this PR, and going to use separate PRs for the rest.

File: llvm/unittests/ProfileData/InstrProfTest.cpp [hunk 1]
```diff
@@ -64,13 +64,130 @@ struct InstrProfTest : ::testing::Test {
...
+      }
+    }
+  }
+
+  // A helper function to test the writes and reads of indirect call value
+  // profiles. The profile writer will scale counters by `ProfWeight` when
+  // adding a record. `Endianness` specifies the endianness used by profile
+  // writer and reader when handling value profile records.
+  void testICallDataReadWrite(
+      uint64_t ProfWeight,
```
Comment (david-xl) [inline llvm/unittests/ProfileData/InstrProfTest.cpp:153]: Can set default weight to 1.

File: llvm/unittests/ProfileData/InstrProfTest.cpp [hunk 2]
```diff
@@ -64,13 +64,130 @@ struct InstrProfTest : ::testing::Test {
...
+        EXPECT_EQ(WantTotalC * ProfWeight, GetTotalC);
+      }
+    }
+  }
+
+  // A helper function to test the writes and reads of indirect call value
+  // profiles. The profile writer will scale counters by `ProfWeight` when
+  // adding a record. `Endianness` specifies the endianness used by profile
+  // writer and reader when handling value profile records.
+  void testICallDataReadWrite(
```
Comment (david-xl) [inline llvm/unittests/ProfileData/InstrProfTest.cpp:152]: Can this setup method be shared with other value profile kind?

File: llvm/unittests/ProfileData/InstrProfTest.cpp (hunk 1 above)
Comment (minglotus-6) [reply llvm/unittests/ProfileData/InstrProfTest.cpp]: done.

File: llvm/unittests/ProfileData/InstrProfTest.cpp (hunk 2 above)
Comment (minglotus-6) [reply llvm/unittests/ProfileData/InstrProfTest.cpp:152]: yes. The plan is to extend this method (rename to 'testICallAndVTableDataReadWrite` and add vtable value profiles) in the patch that introduces vtable as a kind of value profiles.

**Review by david-xl — APPROVED** [review]

File: llvm/unittests/ProfileData/InstrProfTest.cpp [hunk 3]
```diff
@@ -64,13 +64,130 @@ struct InstrProfTest : ::testing::Test {
...
+    Writer.setValueProfDataEndianness(Endianness);
+
+    auto Profile = Writer.writeBuffer();
+    readProfile(std::move(Profile));
+
+    // Set reader endianness.
+    Reader->setValueProfDataEndianness(Endianness);
+
+    Expected<InstrProfRecord> R = Reader->getInstrProfRecord("caller", 0x1234);
+    EXPECT_THAT_ERROR(R.takeError(), Succeeded());
```
Comment (snehasish) [inline llvm/unittests/ProfileData/InstrProfTest.cpp:185]: I think this should be an assertion since we shouldn't access the contents of R if this fails.

File: llvm/unittests/ProfileData/InstrProfTest.cpp [hunk 4]
```diff
@@ -780,94 +845,15 @@ TEST_P(MaybeSparseInstrProfTest, annotate_vp_data) {
...
-  ASSERT_EQ(StringRef((const char *)VD[0].Value, 7), StringRef("callee3"));
-  ASSERT_EQ(StringRef((const char *)VD[1].Value, 7), StringRef("callee2"));
-  ASSERT_EQ(StringRef((const char *)VD[2].Value, 7), StringRef("callee1"));
-
-  // Restore little endian default:
+TEST_P(MaybeSparseInstrProfTest, icall_data_read_write_big_endian) {
+  testICallDataReadWrite(1 /* ProfWeight */, llvm::endianness::big);
+  // Restore little endianness after this test case.
   Writer.setValueProfDataEndianness(llvm::endianness::little);
+  Reader->setValueProfDataEndianness(llvm::endianness::little);
```
Comment (snehasish) [inline llvm/unittests/ProfileData/InstrProfTest.cpp:856]: Why is this needed now? It makes me think that there is some dependency across the tests..

File: llvm/unittests/ProfileData/InstrProfTest.cpp [hunk 5]
```diff
@@ -64,13 +64,130 @@ struct InstrProfTest : ::testing::Test {
...
+    }
+  }
+
+  // A helper function to test the writes and reads of indirect call value
+  // profiles. The profile writer will scale counters by `ProfWeight` when
+  // adding a record. `Endianness` specifies the endianness used by profile
+  // writer and reader when handling value profile records.
+  void testICallDataReadWrite(
+      uint64_t ProfWeight = 1,
+      llvm::endianness Endianness = llvm::endianness::little) {
```
Comment (snehasish) [inline llvm/unittests/ProfileData/InstrProfTest.cpp:154]: I would like to keep the test assertions/expectations in the test themselves, using this helper makes the tests harder to debug on failure. Here is a slightly different refactoring you can consider: These are already parameterized tests so lets extend the params to include endianness and weights. We can combine the params in INSTANTIATE_TEST_SUITE_P using
```
test::Combine(testing::Bool(), 
  Values(TestParam{/*endian=*/little, /*weight=*/1}, 
               TestParam{/*endian=*/little, /*weight=*/10}, 
               TestParam{/*endian=*/big, /*weight=*/1}
  )
)
```
Here TestParam is a new struct which is introduced. This will keep the existing coverage. You can also increase coverage by combining all cases (as cartesian product).

```
test::Combine(testing::Bool(), // sparse?
    Values(little, big), // endianness
    Values(1, 10), // weights
  )
)
```
What do you think? Feel free to avoid if this is requires more time than you want to invest.

File: llvm/unittests/ProfileData/InstrProfTest.cpp [hunk 6]
```diff
@@ -64,13 +64,130 @@ struct InstrProfTest : ::testing::Test {
...
+    std::vector<InstrProfValueData> FuncVD0 = {
+        {uint64_t(callee1), 1}, {uint64_t(callee2), 2}, {uint64_t(callee3), 3}};
+    std::vector<InstrProfValueData> FuncVD2 = {{uint64_t(callee1), 1},
+                                               {uint64_t(callee2), 2}};
+    std::vector<InstrProfValueData> FuncVD3 = {{uint64_t(callee1), 1}};
+    VDArray FuncVD = {FuncVD0, {}, FuncVD2, FuncVD3};
+    addValueProfData(Record1, IPVK_IndirectCallTarget, FuncVD);
+
+    if (ProfWeight == 1U) {
+      Writer.addRecord(std::move(Record1), Err);
```
Comment (snehasish) [inline llvm/unittests/ProfileData/InstrProfTest.cpp:167]: `addRecord` without a weight is a thin convenience wrapper, I think it's ok to drop this if condition in favour of directly passing ProfWeight.

File: llvm/unittests/ProfileData/InstrProfTest.cpp (hunk 5 above)
Comment (minglotus-6) [reply llvm/unittests/ProfileData/InstrProfTest.cpp:154]: At a high level, I'm in favor of keeping test assertions/expectations so that the line numbers are obvious when a test fails.

Future parameterizing the test fixture is a good step, but a cartesian product over all existing test cases of `MaybeSparseInstrProfTest` would create more test cases for all combinations that become vague to reason about (i.e., `value_prof_data_read_write` tests serialization/deserialization of `InstrProfRecord` but invoke profile reader or writer) or  become to adds maintain overhead (i.e., when new aspects are added but not applicable, a test case will do something like `(if std::get<N>(getParam()) == irrelevant) return early`.) 

Since it would probably require more work to refactor the unittest, could I parallel track the test coverage of new feature https://github.com/llvm/llvm-project/pull/66825 and the improvement of unit tests? This way it's less work to keep up with upstream main.

File: llvm/unittests/ProfileData/InstrProfTest.cpp (hunk 4 above)
Comment (minglotus-6) [reply llvm/unittests/ProfileData/InstrProfTest.cpp:856]: The previous test case doesn't have it and the whole unit test passes. I'm adding this line mainly to reset the state but tbh I didn't dig into how this passed previously.

File: llvm/unittests/ProfileData/InstrProfTest.cpp (hunk 5 above)
Comment (snehasish) [reply llvm/unittests/ProfileData/InstrProfTest.cpp:154]: Yes, I am also a little wary of what we might find if we do a cartesian product :)
I'm ok to defer the refactoring so that we can account for the new vtable prof test too. Thanks!

File: llvm/unittests/ProfileData/InstrProfTest.cpp (hunk 6 above)
Comment (minglotus-6) [reply llvm/unittests/ProfileData/InstrProfTest.cpp:167]: yep! This `if-else` is for the test coverage of the thin-wrapper. I could do this in the further parameterized version later.

Github doesn't support re-open of a merged PR so I'll create new PRs for the test.

File: llvm/unittests/ProfileData/InstrProfTest.cpp (hunk 5 above)
Comment (minglotus-6) [reply llvm/unittests/ProfileData/InstrProfTest.cpp:154]: As chatted offline, I realized test coverage for a new kind of value-type might double the amount of code without factoring out common code once I attempted to add new tests. Thanks for going over the test cases together and discussing about how to properly clean it up and prepare for the new value type. I'm sending out smaller patches.
